### PR TITLE
Custom Objection System

### PR DESCRIPTION
**Pull request by cents02**
Features:

- **Custom objections:**
How to enable them:
Simply create a sub folder in a character called "custom_objections" and drop whatever you desire in there as long as the wav name and gif name match.  
**Note: ** Make sure you have a default custom objection in place (_e.g having custom.gif and custom.wav on the chars folder_) in order to have a fallback and not have missingos.

**More important note** Please update your servers as this system won't be useable otherwise.

How to use:
Just right click on the custom objection button, select the objection desired and shoot!

- **/clearocc**

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -66,7 +66,7 @@ build linux x86_64:
       - bin/
 
 build windows i686:
-  image: ${CI_REGISTRY_IMAGE}/builder-windows-i686
+  image: registry.gitlab.com/cents02/ao2-client
```
Comment (oldmud0) [inline .gitlab-ci.yml]: Please revert.

File: include/courtroom.h [hunk 2]
```diff
@@ -479,6 +481,8 @@ class Courtroom : public QMainWindow
   AOButton *ui_realization;
   AOButton *ui_mute;
 
+  QMenu *Custom_obj_menu;
```
Comment (oldmud0) [inline include/courtroom.h]:
```suggestion
  QMenu *custom_obj_menu;
```

File: include/courtroom.h [hunk 3]
```diff
@@ -603,6 +607,7 @@ private slots:
   void on_objection_clicked();
   void on_take_that_clicked();
   void on_custom_objection_clicked();
+  void ShowContextMenu(const QPoint& pos);
```
Comment (oldmud0) [inline include/courtroom.h]:
```suggestion
  void show_context_menu(const QPoint& pos);
```

File: src/courtroom.cpp [hunk 4]
```diff
@@ -1339,8 +1354,14 @@ void Courtroom::handle_chatmessage(QStringList *p_contents)
   append_ic_text(": " + m_chatmessage[MESSAGE], f_showname);
 
   previous_ic_message = f_message;
-
-  int objection_mod = m_chatmessage[OBJECTION_MOD].toInt();
+  bool ok;
+  int objection_mod = m_chatmessage[OBJECTION_MOD].toInt(&ok, 10); //checks if its a custom obj.
+  objection_custom = ""; //added this for safety reasons
+  if(ok == false)
```
Comment (oldmud0) [inline src/courtroom.cpp]:
```suggestion
  if (!ok)
```

File: src/courtroom.cpp [hunk 5]
```diff
@@ -1339,8 +1354,14 @@ void Courtroom::handle_chatmessage(QStringList *p_contents)
...
+  {
+      objection_mod = 4;
+      objection_custom = m_chatmessage[OBJECTION_MOD].split('&')[1]; //takes the name of custom objection
```
Comment (oldmud0) [inline src/courtroom.cpp]: May suffer out-of-bounds error, causing a crash.

File: src/courtroom.cpp [hunk 6]
```diff
@@ -1363,8 +1384,16 @@ void Courtroom::handle_chatmessage(QStringList *p_contents)
       break;
     //case 4 is AO2 only
     case 4:
-      ui_vp_objection->play("custom", f_char, f_custom_theme);
-      objection_player->play("custom.wav", f_char, f_custom_theme);
+      if(objection_custom != "")
+      {
+          ui_vp_objection->play("custom_objections/" + objection_custom, f_char, f_custom_theme);
+          objection_player->play("custom_objections/" + objection_custom.split('.')[0] + ".wav", f_char, f_custom_theme);
```
Comment (oldmud0) [inline src/courtroom.cpp]: We are supposed to be supporting .ogg and .opus as well.

File: src/courtroom.cpp [hunk 7]
```diff
@@ -2939,7 +2968,11 @@ void Courtroom::on_ooc_return_pressed()
       return;
 
   }
-
+  else if(ooc_message.startsWith("/clearooc"))
+  {
+      ui_server_chatlog->clear();
```
Comment (oldmud0) [inline src/courtroom.cpp]: This is not enough. See the data structure that is used for exporting logs.

**Comment by cents02** [comment]
Due to bugs popping up in testing sessions I will close for now

File: src/courtroom.cpp (hunk 6 above)
Comment (cents02) [reply src/courtroom.cpp]: If that's the case I will make so the objection_player searches automatically for all supported formats

File: src/courtroom.cpp (hunk 5 above)
Comment (cents02) [reply src/courtroom.cpp]: For which array are you talking about

Comment (oldmud0) [reply src/courtroom.cpp]: The expression `m_chatmessage[OBJECTION_MOD].split('&')[1]`

Comment (cents02) [reply src/courtroom.cpp]: The only scenarios which will cause out of bound errors are from either somebody modifying the server so there is no packet checking (one of the reasons `If '4'  in button and not '<and>' in button` exists) and somebody modified the client so it doesn't send '&' when appropriate

Comment (oldmud0) [reply src/courtroom.cpp]: Incoming network packets generally should not be trusted. The server sanitizing network input is an expectation, but not a guarantee.

File: include/courtroom.h (hunk 3 above)
Comment (cents02) [reply include/courtroom.h]: This is a qt function so I can't change the name

File: src/courtroom.cpp (hunk 7 above)
Comment (cents02) [reply src/courtroom.cpp]: After exporting the logs it does the same command. Not sure what you mean.

File: include/courtroom.h (hunk 3 above)
Comment (oldmud0) [reply include/courtroom.h]: No. Remember that you `connect` it to the function name you want.

**Pull request closed without merging**